// vue/src/compiler/create-compiler.js
/* @flow */

import { extend } from 'shared/util'
import { detectErrors } from './error-detector'
import { createCompileToFunctionFn } from './to-function'

/**
 * 编译器构造器
 * @param {Function} baseCompile 基础默认的编译构造器
 * @return 返回createCompiler函数，作用创建编译器
 */
export function createCompilerCreator(baseCompile: Function): Function {
  return function createCompiler(baseOptions: CompilerOptions) {
    /**
     * 声明编译器的构造器
     * @param {*} template 
     * @param {*} options 
     * @return 返回baseCompile函数的执行结果——默认编译器对象
     *         数据结构如下：
     *         {
     *           ast,
     *           render: code.render,
     *           staticRenderFns: code.staticRenderFns,
     *           errors,
     *           tips 
     *         }   
     */
    function compile(
      template: string,
      options?: CompilerOptions
    ): CompiledResult {
      const finalOptions = Object.create(baseOptions)
      const errors = []
      const tips = []

      let warn = (msg, range, tip) => {
        (tip ? tips : errors).push(msg)
      }

      if (options) {
        if (process.env.NODE_ENV !== 'production' && options.outputSourceRange) {
          // $flow-disable-line
          const leadingSpaceLength = template.match(/^\s*/)[0].length

          warn = (msg, range, tip) => {
            const data: WarningMessage = { msg }
            if (range) {
              if (range.start != null) {
                data.start = range.start + leadingSpaceLength
              }
              if (range.end != null) {
                data.end = range.end + leadingSpaceLength
              }
            }
            (tip ? tips : errors).push(data)
          }
        }
        // merge custom modules
        if (options.modules) {
          finalOptions.modules =
            (baseOptions.modules || []).concat(options.modules)
        }
        // merge custom directives
        if (options.directives) {
          finalOptions.directives = extend(
            Object.create(baseOptions.directives || null),
            options.directives
          )
        }
        // copy other options
        for (const key in options) {
          if (key !== 'modules' && key !== 'directives') {
            finalOptions[key] = options[key]
          }
        }
      }

      finalOptions.warn = warn

      /**
       * baseCompile返回数据结构
       * {
       *   ast,
       *   render: code.render,
       *   staticRenderFns: code.staticRenderFns
       * }       
       */
      const compiled = baseCompile(template.trim(), finalOptions)
      if (process.env.NODE_ENV !== 'production') {
        detectErrors(compiled.ast, warn)
      }
      compiled.errors = errors
      compiled.tips = tips
      return compiled
    }

    return {
      compile, // function
      compileToFunctions: createCompileToFunctionFn(compile)
    }
  }
}
