// vue/src/core/vdom/helpers/normalize-children.js
/* @flow */

// 引入VNode类，和createEmptyVNode方法
import VNode, { createTextVNode } from 'core/vdom/vnode'
import { isFalse, isTrue, isDef, isUndef, isPrimitive } from 'shared/util'

// The template compiler attempts to minimize the need for normalization by
// statically analyzing the template at compile time.
// 模板编译器试图通过在编译时静态分析模板来最小化规范化的需要。
//
// For plain HTML markup, normalization can be completely skipped
// 对于普通HTML标记，可以完全跳过规范化
// because the generated render function is guaranteed to return Array<VNode>. 
// 因为生成的render函数保证返回Array<VNode>。
// There are two cases where extra normalization is needed:
// 有两种情况需要额外的规范化：

// 1. When the children contains components - 
// 1. 当子级包含组件时
// because a functional component may return an Array instead of a single root. 
// 因为函数组件可能返回一个数组而不是一个根。
// In this case, just a simple normalization is needed -
// 在这种情况下，只需要简单的规范化-
// if any child is an Array, we flatten the whole thing with Array.prototype.concat. 
// 如果有一个子数组，我们就用Array.prototype.concat将整个数组展平。
// It is guaranteed to be only 1-level deep because functional components already normalize their own children.
// 它保证只有1级深度，因为功能组件已经规范化了它们自己的子级。
export function simpleNormalizeChildren (children: any) {
  for (let i = 0; i < children.length; i++) {
    if (Array.isArray(children[i])) {
      return Array.prototype.concat.apply([], children)
    }
  }
  return children
}

// 2. When the children contains constructs that always generated nested Arrays,
// 2.当子级包含总是生成嵌套数组的构造时，
// e.g. <template>, <slot>, v-for, or when the children is provided by user with hand-written render functions / JSX. 
// e.g. <template>, <slot>, v-for, 或者当子对象由用户提供手写的render函数/JSX时。
// In such cases a full normalization is needed to cater to all possible types of children values.
// 在这种情况下，需要完全规范化，以满足所有可能类型的子级values。
export function normalizeChildren (children: any): ?Array<VNode> {
  return isPrimitive(children)
    ? [createTextVNode(children)]
    : Array.isArray(children)
      ? normalizeArrayChildren(children)
      : undefined
}

function isTextNode (node): boolean {
  return isDef(node) && isDef(node.text) && isFalse(node.isComment)
}

function normalizeArrayChildren (children: any, nestedIndex?: string): Array<VNode> {
  const res = []
  let i, c, lastIndex, last
  for (i = 0; i < children.length; i++) {
    c = children[i]
    if (isUndef(c) || typeof c === 'boolean') continue
    lastIndex = res.length - 1
    last = res[lastIndex]
    //  nested
    if (Array.isArray(c)) {
      if (c.length > 0) {
        c = normalizeArrayChildren(c, `${nestedIndex || ''}_${i}`)
        // merge adjacent text nodes
        if (isTextNode(c[0]) && isTextNode(last)) {
          res[lastIndex] = createTextVNode(last.text + (c[0]: any).text)
          c.shift()
        }
        res.push.apply(res, c)
      }
    } else if (isPrimitive(c)) {
      if (isTextNode(last)) {
        // merge adjacent text nodes
        // this is necessary for SSR hydration because text nodes are
        // essentially merged when rendered to HTML strings
        res[lastIndex] = createTextVNode(last.text + c)
      } else if (c !== '') {
        // convert primitive to vnode
        res.push(createTextVNode(c))
      }
    } else {
      if (isTextNode(c) && isTextNode(last)) {
        // merge adjacent text nodes
        res[lastIndex] = createTextVNode(last.text + c.text)
      } else {
        // default key for nested array children (likely generated by v-for)
        if (isTrue(children._isVList) &&
          isDef(c.tag) &&
          isUndef(c.key) &&
          isDef(nestedIndex)) {
          c.key = `__vlist${nestedIndex}_${i}__`
        }
        res.push(c)
      }
    }
  }
  return res
}
